### ISSUE TITLE
Bug when typing "s"/"d" with another character in rapid succession

### ISSUE THREAD

**Issue by denosawr** [jasonrudolph/keyboard#40]
Steps to reproduce: Type "s" or "d" followed by another character in rapid succession (`< MAX_TIME_BETWEEN_SIMULTANEOUS_KEY_PRESSES`), such as `se` really fast. Instead of inserting `se`, `es` is inserted.
Running on macOS 10.13.5 with Hammerspoon 0.9.66. This issue has been consistently reproduced with Hammerspoon running, and has been consistently absent with Hammerspoon closed.

**Comment by denosawr** [comment]
This issue is resolved by pull request #41, which will merely wait for a key press during the interval in which the program checks for the other key which composes the Super Duper shortcut to be pressed. Afterwards, it inserts both characters in correct succession.

**Comment by denosawr** [comment]
Friendly ping?

**Comment by jasonrudolph** [comment]
@denosawr: Thanks for this contribution! It might be a while before I can properly review it, but it's on my list. 🙂

**Comment by denosawr** [comment]
Awesome, good to know it's been acknowledged!

**Comment by rickharris** [comment]
If there's anything I can do to help with this, I'd be glad to. Been using this for a few years and love it! But every day there's enough out-of-order "s" and "d" characters to make me want to disable super-duper mode.

Is the linked PR a good starting point, or do other options need to be explored?

**Comment by jasonrudolph** [comment]
> Been using this for a few years and love it!

@rickharris: Awesome! I'm so glad to hear that! 💟

> Is the linked PR a good starting point, or do other options need to be explored?

https://github.com/jasonrudolph/keyboard/pull/41 will probably work for you most of the time, but it's susceptible to a race condition as described in https://github.com/jasonrudolph/keyboard/pull/41#issuecomment-406087831. With a super short timeout, you probably won't run into the problems caused by the race condition, but you might.

Due to the potential for the race condition, I'm not comfortable merging https://github.com/jasonrudolph/keyboard/pull/41 in its current state. But if it you want to give it a shot, you can use the fork at https://github.com/denosawr/keyboard or apply 111b1f3555fa5911cfc0c396cada8599d41694f5 to your local clone of https://github.com/jasonrudolph/keyboard.

I hope that context is helpful.

**Comment by denosawr** [comment]
Okay, so setting aside the... uh, issue of me accidentially closing the PR, I have a new commit which seems to fix the race condition, as well as addressing some issues the previous implementation caused (namely inability to use s/d + special keys like `space`). On my end, this PR seems to work.

Note that I have found instances where I type "s" and then "e" and "es" is typed. However, using if you log the calls, it turns out that this happens only when the "e" is typed just after the timeout. I can't really think of a way to address this well, and it should only show up in a very short time window, very rarely. If anyone has any ideas, that'd be awesome! :)

**Comment by acheronfail** [comment]
In order to fix this problem, I moved the trigger out of Hammerspoon and into Karabiner, and it works much, much better in my opinion.

Basically, remove the triggers in super duper mode in Hammerspoon, and expose a way to toggle super duper mode via the command line (see https://www.hammerspoon.org/docs/hs.ipc.html#cliInstall).

I did it by exporting the `superDuperMode` table which contains the `superDuperMode:enter()` and `superDuperMode:reset()` functions:

```lua
-- ... the rest of the super duper module ...

-- Expose superDuperMode so we can control it from other modules (and the CLI).
return superDuperMode
```

Then, add a complex modification like this to karabiner:

```json
{
    "manipulators": [
        {
            "description": "(S)uper (D)uper mode trigger",
            "from": {
                "simultaneous": [
                    {
                        "key_code": "s"
                    },
                    {
                        "key_code": "d"
                    }
                ],
                "simultaneous_options": {
                    "to_after_key_up": [
                        {
                            "shell_command": "/usr/local/bin/hs -c 'superDuperMode:reset()'"
                        }
                    ]
                }
            },
            "parameters": {
                "basic.simultaneous_threshold_milliseconds": 40
            },
            "to": [
                {
                    "repeat": false,
                    "shell_command": "/usr/local/bin/hs -c 'superDuperMode:enter()'"
                }
            ],
            "type": "basic"
        }
    ]
}
```

That will toggle the mode if both `S` and `D` are pressed, and Karabiner is much better at timing the events and making sure everything works as expected.

**Comment by jasonrudolph** [comment]
This is likely resolved by #61, so I'm going to close this issue.

If you're still able to reproduce the problem using the latest code on the `master` branch, please let me know, and I'll re-open this issue. :bow:
